### ISSUE TITLE
[Sandbox] OpenTofu

### ISSUE THREAD

**Issue by pawelpiwosz** [cncf/sandbox#81]
### Application contact emails

core@opentofu.org

### Project Summary

The open source, cloud agnostic, Infrastructure as Code tool

### Project Description

OpenTofu is a Terraform fork created in response to HashiCorp licenses change. The goal of the founders was to keep this tool open source. We believe that IaC and Terraform (now OpenTofu) is a skeleton of Cloud Native applications, and by ensuring the competition and free use of the main IaC tool, we ensure the further development of the whole IaC space.

By giving the possibility to the community to not only participate in the project as developers but also participate in ideas and decisions about the further development, we want to ensure a truly open source and community driven approach, and OpenTofu can respond to changes, needs and expectations of the users.

OpenTofu itself provides the "interpreter" for HCL code and allows to communicate and interact with different cloud providers (but not only) by library of providers. OpenTofu group is not an owner of providers, these are created and maintained by community and / or companies. The collection of providers which can be used with Opentofu is provided by Providers Registry, available under this link: https://registry.opentofu.org/. OpenTofu is also responsible for developing, maintaining and growing this registry.
Providers themselves are not owned by the OpenTofu group. Everyone can write the provider and if the provider is available on GitHub, most likely will be automatically available for OpenTofu and in the Registry. With this approach we decrease the time needed to enable new providers in the landscape virtually to zero.

Additional layer is the modules layer. Modules are sets of templates created and maintained by the community (mostly) and they provide the possibility to use proven templates for specific use cases, like network creation, hardening database setup, etc.

As  described, OpenTofu plays a vital and central part in the IaC landscape, where there are HCL templates on one side, providers on the second and executable modules on the other.

### Org repo URL (provide if all repos under the org are in scope of the application)

https://github.com/opentofu

### Project repo URL in scope of application

https://github.com/opentofu/opentofu

### Additional repos in scope of the application

https://github.com/opentofu/manifesto
https://github.com/opentofu/opentofu.org
https://github.com/opentofu/brand-artifacts
https://github.com/opentofu/registry
https://github.com/opentofu/scripts

### Website URL

https://opentofu.org/

### Roadmap

https://github.com/opentofu/opentofu/milestones

### Roadmap context

The project’s team decided to keep the roadmap as GitHub milestones, which gives full transparency to the process and the current state of the process. 
Additionally, team provides weekly updates to public knowledge: https://github.com/opentofu/opentofu/blob/main/WEEKLY_UPDATES.md 
Roadmap is agreed and confirmed by Steering Committee, and information about decisions is also publicly available: https://github.com/opentofu/opentofu/blob/main/TSC_SUMMARY.md 

### Contributing Guide

https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md

### Code of Conduct (CoC)

https://github.com/opentofu/opentofu/blob/main/CODE_OF_CONDUCT.md

### Contributing or Sponsoring Org

https://opentofu.org/supporters

### Maintainers file

https://github.com/opentofu/opentofu/blob/main/MAINTAINERS

### IP Policy

- [X] If the project is accepted, I agree the project will follow the CNCF IP Policy

### Trademark and accounts

- [X] If the project is accepted, I agree to donate all project trademarks and accounts to the CNCF

### Why CNCF?

We believe that CNCF is a natural place for IaC tools like Opentofu. IaC is Cloud Native by nature, it allows users to explore and enforce all best practices related to infrastructure in cloud native world (like scalability, sustainability, resiliency, immutability, idempotency, security, accessibility, and so on).
Our goal is to ensure that OpenTofu is and will be a truly open source project.

OpenTofu can contribute to CNCF in number of ways, including
* Maintain and develop core element of OpenTofu toolset - the tofu binary. This is the central part of the project, as tofu interprets the HCL language and uses providers to deliver templates to the cloud.
* Maintain and develop OpenTofu Registry, to achieve easy and seamless options to provide and use different providers to the community.
* Build an open source community around the Infrastructure as Code principles.
* Growing the Cloud Native ecosystem by providing first-class toolset to deploy infrastructure.

We are looking forward to being part of CNCF, to be part of the big family of initiatives which makes cloud adoption faster, more reliable and more secure. 
For the project it is a chance to:
* Access to a big community of contributors. This might help the project to grow and maintain the functionalities. Infrastructure as code is vital and integral part of cloud native landscape. Possibility to work with an active and innovative community is a confirmation of opportunities to develop the product in expected direction.
* The main contributors are well established on the market, but with help from CNCF in marketing, funding and training, we together can ensure the progress and quick response to change in the landscape.
* As a CNCF project, the visibility and adoption of OpenTofu will grow, as well as trust in the project and its goals.

### Benefit to the Landscape

OpenTofu greatly fits the landscape by providing the tool which uses the most popular infrastructure as code language model - HCL. According to different reports the market share of Terraform based IaC is around 80%. It shows the potential for HCL based tools.

Our project is the one which can bind other CNCF projects together by providing create, configure, manage and update capabilities for the underlying infrastructure needed for. 

OpenTofu might be an enabler for multiple projects, providing the tool to provision infrastructure. This is very important for CNCF projects, as CNCF mentioned that projects should search for alternatives to Terraform.

### Cloud Native 'Fit'

OpenTofu is an infrastructure as code tool. Infrastructure as code approach embodied all cloud native pillars, especially immutable infrastructure and declarative approach. Infrastructure as code is an enabler for automation, resiliency, knowledge management and capacity management.
This project fits into the Automation / Configuration category in the CNCF landscape. We see the project under App delivery TAG. As automation for IaC tool, OpenTofu enables automation of infrastructure for many projects which already exist in the CNCF landscape.

### Cloud Native 'Integration'

All projects under CNCF and all projects in the CNCF landscape can benefit from using OpenTofu as their tool for automation of the infrastructure. OpenTofu doesn’t depend on any other project under the CNCF umbrella.

OpenTofu might be an enabler for multiple projects, providing the tool to provision infrastructure. This is very important for CNCF projects, as CNCF mentioned that projects should search for alternatives to Terraform.

### Cloud Native Overlap

No overlap. The closest is Crossplane, however, while Crossplane is a control plane to manage application and infrastructure, OpenTofu is an ecosystem to declaratively build infrastructure.

### Similar projects

There are no similar projects under the CNCF umbrella. The closest might be Crossplane, as the project related to Infrastructure as Code.

There is no overlap, though. Crossplane is a control plane for application and infrastructure management, while OpenTofu is an ecosystem providing declarative ways of defining and managing infrastructure.

### Landscape

OpenTofu is not listed in the landscape. Part of the initiative leaders is present on the landscape (Env0, Harness, Spacelift)

### Business Product or Service to Project separation

All initiative leaders are owners of platforms and/or tools to manage infrastructure as code. This gives us a unique understanding of the area as well as industry. In the OpenTofu manifesto (https://opentofu.org/manifesto) we articulated our goals for this project, to ensure healthy market collaboration and competition. 

Number of entities engaged with their support and people and huge support from the community ensures the diversity, separation and independence of the project. Being part of the Linux Foundation shows our dedication to open-source and community.

The initiative leaders pledged their resources and efforts to maintain the project. Two aspects here are most important:
* Hire people solely for OpenTofu. A few open source engineers were already hired by the leaders
* Ensure that OpenTofu is and will be an open source product. Currently OpenTofu is under Linux Foundation. As explained in "why CNCF" and "Cloud Native fit", we believe that CNCF is a natural place for tools like OpenTofu.

### Additional information

Important note: OpenTofu contributions and releases are made under the Mozilla Public License, version 2.0 (MPL-2.0). OpenTofu started as a fork of last open source release of Terraform, which was licensed as MPL-2.0. The OpenTofu license is specified in the repository, at: https://github.com/opentofu/opentofu/blob/main/LICENSE 

As a fork of Terraform's MPL-2.0 codebase, we do not have an option to relicense the codebase. This license remains consistent for everyone who took a dependency on the Terraform project when it was open source.

Mozilla Public License 2.0 (MPL 2.0) is a simple copyleft license. The MPL's "file-level" copyleft is designed to encourage contributors to share modifications they make to the code, while still allowing them to combine the code with code under other licenses (open or closed/proprietary) with minimal restrictions. We believe the MPL 2.0 license is suitable for the intended purpose and will contribute to the project’s success.

**Comment by caniszczyk** [comment]
Thanks for your submission! This will require either IP changes in CNCF to allow a MPL-2.0 project only
https://github.com/cncf/foundation/blob/main/charter.md#11-ip-policy

@amye can you ensure this is put on the IP Committee's agenda at an upcoming meeting.

**Comment by TheFoxAtWork** [comment]
Moving this application to postponed due to outstanding issues requiring resolution before the TOC can discuss this project's application. CC @caniszczyk

**Comment by caniszczyk** [comment]
FYI the TOC would like to wait until https://opentofu.org/blog/our-response-to-hashicorps-cease-and-desist/ fully resolves for more detail

**Comment by mrbobbytables** [comment]
@pawelpiwosz I know there's been a lot that has happened since April; just wanting to check in and see if you had an update on the current state of things.^^;;

**Comment by mrbobbytables** [comment]
> pawelpiwosz I know there's been a lot that has happened since April; just wanting to check in and see if you had an update on the current state of things.^^;;

@cube2222 is this something you could potentially follow up on?

**Comment by cube2222** [comment]
Hey @mrbobbytables,
other than the C&D and OpenTofu’s response to it, HashiCorp also filed a [DMCA takedown request](https://github.com/github/dmca/blob/master/2024/04/2024-04-11-hashicorp.md), which OpenTofu responded to with a [DMCA counter notice](https://github.com/github/dmca/blob/master/2024/04/2024-04-24-opentofu-counternotice.md). Since then, we’ve been informed by GitHub that they haven’t received any notice about HashiCorp filing an action seeking a court order to restrain OpenTofu from posting its content.

In other words, we currently consider the matter closed.

**Comment by kvendingoldo** [comment]
OpenTofu can't be a part of CNCF, due to rules violation: https://github.com/opentofu/registry/pull/824

**Comment by he0s** [comment]
I'm supporting @kvendingoldo as soon as the OpenTofu manifesto [includes](https://github.com/opentofu/manifesto/commit/4d2fcfe47711b18481cadef836c5c4f0ed992c43#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L3) the following statement, which is violated by its developers or some core team members in [#817](https://github.com/opentofu/registry/pull/817):

```
We urge HashiCorp to reconsider and switch Terraform back to an open source license, avoiding fragmentation of the community.
```
It obviously brings fragmentation of the community and assumes some new HashiCorp Terraform or OpenTofu forks will appear as response to the harmful actions from the OpenTofy team.

**Comment by cam72cam** [comment]
Hi, OpenTofu Tech Lead here speaking on behalf of the Technical Steering Committee.

We have taken proactive steps to ensure compliance with relevant sanctions. We have heard the community’s feedback on this issue and are working on a more detailed response and next steps. We will update the community soon.

**Comment by kvendingoldo** [comment]
@mkdolan Do you have any news about exclusion?

**Comment by cam72cam** [comment]
In discussing with the Linux Foundation (including @mkdolan), we understand this is a complex issue with many countries imposing sanctions on Russia and we do not have the community resources to figure this all out right now. Anyone in any geography could provide a similar service locally.

**Comment by tsunamaru** [comment]
@cam72cam what the freak are you trying to "figure out" here? It is clearly stated by Linux Foundation that [public OSS is not a subject for export controls](https://www.linuxfoundation.org/resources/publications/understanding-us-export-controls-with-open-source-projects), period.

> Anyone in any geography could provide a similar service locally.

Remove that "avoiding fragmentation of the community" shit from your manifesto then, you're not better than HashiCorp itself.

**Comment by cam72cam** [comment]
That publication was last "updated July 2021".  We are working with the Linux Foundation directly to address this.

**Comment by kvendingoldo** [comment]
@cam72cam until it won't be updated by Linux Foundation (hope it won't) the project should follow these rules, isn't it true?

**Comment by cam72cam** [comment]
We are taking actions based on the current and direct advice we are receiving from the Linux Foundation.  It is a complex situation and we will strive to provide clarity where we can.

**Comment by he0s** [comment]
> Anyone in any geography could provide a similar service locally.

That's true, but this statement leads to a non-inclusive community you're trying to build around the service that founded as response to the changes the initial project (Terraform) did when they changed their licensing model.

What is disappointing me in this case that the Linux Foundation doesn't provide any information publicly and all conversations and decision "happening" around the case and the project are private, like in enterprises that work only for money, not for community and their users.

This is a sign that a new Terraform fork needs to be created, that will follow its manifesto and put there openness, inclusiveness and similar rules as a ground to make the project for real community that doesn't include any peace of politics into the project.

**Comment by slonopotamus** [comment]
I thought the issue that triggered lots of people was not the service, but removal of code.

**Comment by TheFoxAtWork** [comment]
Community - 
  This issue is for the application of OpenTofu into the CNCF. It was been placed on hold from review by the TOC ([refer to previous comment](https://github.com/cncf/sandbox/issues/81#issuecomment-2085529676)).

  The TOC is aware of the recent changes to the OpenTofu project and has ensured CNCF legal is aware, and we've been informed they are in discussion with head of LF legal regarding those changes (who is subsequently in communication with the OpenTofu maintainers).

  We understand the community has many open questions regarding the current status, guidance, and direction regarding the changes to the OpenTofu project, however the project's sandbox application is not the mechanism to field them. We encourage concerned community members to engage with the OpenTofu project on their repo and we ask community members to seek understanding with patience while the individuals involved work through next steps.
  
  Current CNCF projects with questions or seeking guidance on OFAC sanctions and other export control laws should file an issue with the CNCF Service Desk.
  
  Until CNCF staff inform the TOC that previous cease and desist is resolved and that no other issues impeding review are present, the project's will not be reviewed for inclusion.

**Comment by TheFoxAtWork** [comment]
updating the label to "need-info" on this to reflect its true status and moving to the waiting column given the comment on Sept 5th. Please note that the information being requested is from the CNCF, not from the project itself.

**Comment by TheFoxAtWork** [comment]
@caniszczyk @mrbobbytables would you both check on the status of this if anything remains outstanding? We're planning a January Sandbox meeting and i'd like to move this onto the agenda if possible.  Last i recall we were waiting on resolution of the cease and desist the project had been receiving.

**Comment by rawkode** [comment]
Isn't this already a Linux Foundation project? So there's no real value, for end users, in this moving to the CNCF.

**Comment by dims** [comment]
> something that will need to be sorted out if approved is OpenTofu itself is MPL-2.0 and would require GB approval.

@mrbobbytables 11 (d) of the foundation charter says `All outbound code will be made available under the Apache License, Version 2.0.` so it feels like more than approval and sounds like a charter change.

**Comment by jberkus** [comment]
TAG Contributor strategy has reviewed this project and found the following:

* The [contributor guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md) is very complete, including some debugging guides and checklists
* OpenTofu is governed by a [Technical Steering Committee](https://github.com/opentofu/opentofu/blob/main/TSC_SUMMARY.md) but how that team is elected or replaced isn't documented any place I could find, but has something to do with project sponsorship.  There's also a plethora of subprojects, in the form of Providers, and how those are governed is also undefined.   
* The [roadmap](https://github.com/opentofu/opentofu/milestones) consists of Github milestones, and appears up to date and very active
* There are 6 [maintainers](https://github.com/opentofu/opentofu/blob/main/MAINTAINERS), who work for at least two, possibly 3-4 different organizations, including ENV0 and Spacelift.  The maintainers are not the same group as the TSC.  One maintainer has a deleted Github account.

It's likely, given that OpenTofu is already an LF project, that governance documentation exists somewhere.  However, those docs are not public and easily findable.

This review is for the TOC’s information only.  Sandbox projects are not required to have full governance or contributor documentation.

**Comment by dims** [comment]
> > something that will need to be sorted out if approved is OpenTofu itself is MPL-2.0 and would require GB approval.

> @mrbobbytables 11 (d) of the foundation charter says All outbound code will be made available under the Apache License, Version 2.0. so it feels like more than approval and sounds like a charter change.

> https://github.com/cncf/foundation/blob/main/charter.md

@caniszczyk TOC would like to know about this ^^ question.

**Comment by mrbobbytables** [comment]
Okay - After further discussion. We cannot proceed with considering OpenTofu as is, and I am going to remove it from the review queue for now.

Following [Section 11, the IP Policy](https://github.com/cncf/foundation/blob/main/charter.md#11-ip-policy) of the [CNCF Charter](https://github.com/cncf/foundation/blob/main/charter.md), we may only accept Apache 2.0 projects and all outbound code must also be Apache 2.0, however the GB may make an exception on a case by case basis.

> 11. IP Policy
> ...
>
> (c) All new inbound code contributions to the CNCF shall be (i) accompanied by a Developer Certificate of Origin sign-off (https://developercertificate.org/) and (ii) made under the Apache License, Version 2.0 (available at https://www.apache.org/licenses/LICENSE-2.0), such license to be in addition to, and shall not supersede, obligations undertaken under the contribution license agreement(s) provided for in (b) above.
>
> (d) All outbound code will be made available under the Apache License, Version 2.0.
>
> (e) All projects evaluated for inclusion in the CNCF shall be completely licensed under an OSI-approved open source license. If the license for a project included in CNCF is not Apache License, Version 2.0, approval of the Governing Board shall be required.


OpenTofu can:
- relicense under Apache 2.0
- Follow up with the CNCF Governing Board to seek an exception for the project or the MPL 2.0 license itself.

**Comment by rawkode** [comment]
OpenTofu is a LF project, why the lateral movement? It seems there is no value here for the community.

**Comment by openuuuser** [comment]
@rawkode maybe because opentofu is not open?

https://github.com/opentofu/registry/pull/818
https://github.com/opentofu/registry/issues/2008
https://github.com/opentofu/registry/issues/937
https://github.com/opentofu/registry/issues/68
https://github.com/opentofu/registry/issues/23

**Comment by stevehipwell** [comment]
> OpenTofu is a LF project, why the lateral movement? It seems there is no value here for the community.

@rawkode I'm interested in your thought process here if you have more to add?

**Comment by rawkode** [comment]
If we look at why developers and organisations adopt open-source projects, it's because they are governed by a foundation to provides some assurances around neutrality and license rug-pulls.

The community already has these assurances on OpenTofu from its home in the Linux Foundation.

Based on conversations from KubeCon, it looks like OpenTofu aren't getting the adoption they expected, nor the contributions; I suspect this is the true motive for the lateral movement.

I'm not saying that is a bad reason for them to seek CNCF support; but it should be transparent to the TOC as they vote.

**Comment by orbitz** [comment]
> Based on conversations from KubeCon, it looks like OpenTofu aren't getting the adoption they expected, nor the contributions; I suspect this is the true motive for the lateral movement.

I do not know what conversations you had at KubeCon but this doesn't seem like the correct forum for speculation.

> I'm not saying that is a bad reason for them to seek CNCF support; but it should be transparent to the TOC as they vote.

You've gone from suspecting a reason for the move to CNCF to now this is information that people should be aware of when they vote?

**Comment by marcinwyszynski** [comment]
@rawkode It's rather unusual for someone to base their argument on entirely unverifiable gossip on a public forum using their real name. But let's not focus on your motivations, let's focus on facts.

There is no "lateral movement" here - just look at the date of the original application! Four days after our first release.

A bit of history then. When OpenTofu was formed, multiple open source foundations approached us offering to host the project. We believed that as an important technology underpinning many cloud tools it belonged with the likes of Kubernetes, Open Policy Agent or Jaeger. However, CNCF only accepts Apache-licensed projects, and we needed to seek an exception for MPL 2.0 before we were even eligible to join. We are still being hosted by CNCF's parent foundation - LF - which does not have such restrictions. So no, it's not a lateral movement, not something we've just started.

And look, it's not like the success of the project depends on joining CNCF, and being voted against spells the end of OpenTofu. From day one we thought it was a good fit but if the decision makers think otherwise, so be it. We are making the project great by shipping features, fixing bugs and engaging with the community, and letting our actions speak for us.

**Comment by kvendingoldo** [comment]
Unfortunately, I’m unable to add a reaction to @marcinwyszynski’s comment, but I do agree with his point. 

Success doesn’t hinge on joining CNCF. OpenTofu is an excellent project, and the maintainers are making efforts to avoid further fragmentation within the community, which was a concern after some events of last September. 

Additionally, OpenTofu has already become a strong alternative to Terraform, giving users the flexibility to choose tools based on their specific requirements. Based on that, it has to be a part of CNCF.

**Comment by mrbobbytables** [comment]
the git-vote bot runs on a cronjob, so it hasn't chimed in that the vote has passed yet, but going by /check-vote status we are good! The vote has passed.

I've created an onboarding issue that we can use to track the migration and continue any further discussion there. :)
https://github.com/cncf/sandbox/issues/367
